kernel/of/fdt.c

```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+        node = fdt_next_node(fdt, node, NULL);
     }
 
     return NULL;
 }
+
+struct fdt_node*
+fdt_find_node_by_device_type(struct fdt* fdt, struct fdt_node* start, const char* type)
+{
+    struct fdt_node* node;
+    if (start != NULL) {
+        node = fdt_next_node(fdt, start, NULL);
+    }
+    else {
+        node = fdt_node_begin(fdt);
+    }
+
+    size_t len = (size_t)strlen(type);
+    while (node != NULL) {
+        struct fdt_prop* prop = fdt_get_prop_by_name(fdt, node, NULL, "device_type");
+        if (prop != NULL) {
+            size_t prop_len = fdt_prop_val_len(prop);
+            void* prop_val = fdt_prop_val(prop);
+            // Consider equal even if '\0' is omitted in the property
+            if (prop_len == len + 1 || prop_len == len) {
+                if (memcmp(type, prop_val, prop_len) == 0) {
+                    return node;
+                }
+            }
+        }
+        node = fdt_next_node(fdt, node, NULL);
+    }
+
+    return NULL;
+}
+
+struct fdt_node*
+fdt_find_node_by_unit_name(struct fdt* fdt, struct fdt_node* start, const char* name)
+{
+    struct fdt_node* node;
+    if (start != NULL) {
+        node = fdt_next_node(fdt, start, NULL);
+    }
+    else {
+        node = fdt_node_begin(fdt);
+    }
+
+    while (node != NULL) {
+        if (strcmp(name, fdt_node_name(node)) == 0) {
+            return node;
+        }
+        node = fdt_next_node(fdt, node, NULL);
+    }
+
+    return NULL;
+}
+
+struct fdt_prop*
+fdt_node_get_inherited_prop(
+    struct fdt* fdt, struct fdt_node* node, const char* prop_name
+)
+{
+    struct fdt_prop* prop = fdt_get_prop_by_name(fdt, node, NULL, prop_name);
+    if (prop != NULL) {
+        return prop;
+    }
+
+    int max_depth = fdt_max_depth(fdt);
+    if (max_depth < 0) {
+        // Corrupted FDT
+        return NULL;
+    }
+
+    struct fdt_node* parents[max_depth];
+
+    size_t num_parents = fdt_node_get_parents(fdt, node, parents, (size_t)max_depth);
+    for (size_t i = 0; i < num_parents; i++) {
+        prop = fdt_get_prop_by_name(fdt, parents[i], NULL, prop_name);
+        if (prop != NULL) {
+            return prop;
+        }
+    }
+
+    return NULL;
+}
+
+uint32_t
+fdt_node_get_address_cells(struct fdt* fdt, struct fdt_node* node)
+{
+    struct fdt_prop* prop = fdt_node_get_inherited_prop(fdt, node, "#address-cells");
+
+    if (prop == NULL) {
+        // Not in the standard but Linux allows it
+        // so people do it.
+        return 2;
+    }
+
+    size_t num_cells = fdt_prop_num_cells(prop);
+    if (num_cells == 0) {
+        // This should never happen (Corrupted FDT)
+        // TODO: log warning
+        return (uint32_t)0;
+    }
+
+    if (num_cells > 1) {
+        // TODO: log a warning (this is unexpected)
+        //                     (not panic worthy though)
+    }
+
+    return fdt_prop_get_cell(prop, 0);
+}
+
+uint32_t
+fdt_node_get_size_cells(struct fdt* fdt, struct fdt_node* node)
```

Comment:
The helper function is "fdt_node_get_inherited_prop", both of these give an example of how to use that function. A helper function for this would have to be named `fdt_node_get_first_cell_of_inherited_prop_which_must_exist_and_must_have_a_single_non_zero_cell` which is not a function I will ever write and I don't think you should either

Reply:
`fdt_node_get_first_cell_of_prop`?

Reply:
That's just `fdt_node_get_cell(node,0)`

Reply:
The rest of the code is just making sure that property has at least 1 cell

Reply:
The verification code could be extracted into a helper

Reply:
I mean yeah, but at a certain point where the extra helper functions hurt code simplicity, don't sacrifice KISS for DRY and vice versa.

Reply:
Ya but there's like 15 LOC for validation

Reply:
It's just "get array", "get array length", "check access is in bounds", "access element". It's an insanely common pattern but it doesn't mean it should be abstracted away. Unless you want every single array access in the kernel to go through a helper function.

Reply:
Just leave it